src/Compilers/Core/Portable/SourceGeneration/IncrementalContexts.cs

```diff
@@ -198,6 +207,40 @@ internal GeneratorFilterContext(ISourceGenerator generator, CancellationToken ca
         public CancellationToken CancellationToken { get; }
     }
 
+    /// <summary>
+    /// Context passed to an incremental generator when it has registered an output via <see cref="IncrementalGeneratorInitializationContext.RegisterHostOutput{TSource}(IncrementalValuesProvider{TSource}, Action{HostOutputProductionContext, TSource})"/>
+    /// </summary>
+    [Experimental(RoslynExperiments.GeneratorHostOutputs, UrlFormat = RoslynExperiments.GeneratorHostOutputs_Url)]
+    public readonly struct HostOutputProductionContext
+    {
+        internal readonly ArrayBuilder<(string, object)> Outputs;
+
+        internal HostOutputProductionContext(ArrayBuilder<(string, object)> outputs, CancellationToken cancellationToken)
+        {
+            Outputs = outputs;
+            CancellationToken = cancellationToken;
+        }
+
+        /// <summary>
+        /// Adds a host specific output
+        /// </summary>
+        /// <param name="name">The name of the output to be added.</param>
+        /// <param name="value">The output to be added.</param>
+        /// <remarks>
+        /// A host output has no defined use. It does not contribute to the final compilation in any way. Any outputs registered 
+        /// here are made available via the <see cref="GeneratorRunResult.HostOutputs"/> collection, and it is up the host to
+        /// decide how to use them. A host may also disable these outputs altogether if they are not needed. The generator driver
+        /// otherwise makes no guarantees about how the outputs are used, other than that they will be present if the host has
+        /// requested they be produced.
+        /// </remarks>
+        public void AddOutput(string name, object value) => Outputs.Add((name, value));
```

Comment:
Shouldn't this also be a dictionary so we throw exceptions for duplicate names early?

Reply:
This one is tricky. Because of the way the NodeStateTables work, we would need to share the dictionary across multiple calls of the callback. It's not impossible, and in fact I actually started doing it, but it's not simple.

This is the same way `RegisterSourceOutput` works today, we effectively let you add duplicates then throw when we combine them (which was different from V1 generators where it was at the point of addition). I agree it's a little odd, but you do get the same effect, just a different call stack, and the complexity of it didn't seem worth doing.